panels/sharing/cc-sharing-panel.c

```diff
@@ -661,34 +661,69 @@ cc_sharing_panel_setup_media_sharing_dialog (CcSharingPanel *self)
 }
 
 static void
-content_sharing_switch_active_changed_cb (GtkSwitch *widget,
-                                          GParamSpec *pspec,
-                                          CcSharingPanel *self)
+enter_discoverable_mode_done_callback (GObject      *proxy,
+                                       GAsyncResult *result,
+                                       gpointer      user_data)
 {
+  CcSharingPanel *self = user_data;
   GError *error = NULL;
+  GVariant *return_tuple = g_dbus_proxy_call_finish (G_DBUS_PROXY (proxy), result, &error);
 
-  if (gtk_switch_get_active (widget))
+  if (!return_tuple)
     {
-      g_dbus_proxy_call_sync (self->priv->companion_app_avahi_helper_proxy,
-                              "EnterDiscoverableMode",
-                              g_variant_new ("(s)", "Endless Computer"),
-                              G_DBUS_CALL_FLAGS_NONE, -1,
-                              NULL, &error);
-    }
-  else
-    {
-      g_dbus_proxy_call_sync (self->priv->companion_app_avahi_helper_proxy,
-                              "ExitDiscoverableMode",
-                              NULL,
-                              G_DBUS_CALL_FLAGS_NONE, -1,
-                              NULL, &error);
+      g_message ("Error when calling EnterDiscoverableMode on Companion App Helper Proxy: %s",
+                 error->message);
+      g_clear_error (&error);
+      return;
     }
 
-  if (error != NULL)
+  g_variant_unref (return_tuple);
+  gtk_switch_set_state (GTK_SWITCH (self->priv->content_sharing_switch), TRUE);
+}
+
+static void
+exit_discoverable_mode_done_callback (GObject      *proxy,
+                                      GAsyncResult *result,
+                                      gpointer      user_data)
+{
+  CcSharingPanel *self = user_data;
+  GError *error = NULL;
+  GVariant *return_tuple = g_dbus_proxy_call_finish (G_DBUS_PROXY (proxy), result, &error);
+
+  if (!return_tuple)
     {
-      g_critical ("Unable to set the enabled state of content sharing: %s", error->message);
-      g_error_free (error);
+      g_message ("Error when calling ExiDiscoverableMode on Companion App Helper Proxy: %s",
+                 error->message);
+      g_clear_error (&error);
+      return;
     }
+
+  g_variant_unref (return_tuple);
+  gtk_switch_set_state (GTK_SWITCH (self->priv->content_sharing_switch), FALSE);
+}
+
+
+static void
```

Comment:
The `state-set` callback has a `gboolean` return parameter, and this function should return `TRUE` to use the delayed state change machinery.

Reply:
Of course, I forgot about that.